addons/web/static/src/views/fields/section_note_list/section_note_list_field.js

```diff
@@ -0,0 +1,387 @@
+import { _t } from "@web/core/l10n/translation";
+import { x2ManyCommands } from "@web/core/orm_service";
+import { registry } from "@web/core/registry";
+import { mergeClasses } from "@web/core/utils/classname";
+import { getId } from "@web/model/relational_model/utils";
+import { x2ManyField, X2ManyField } from "@web/views/fields/x2many/x2many_field";
+import { ListRenderer } from "@web/views/list/list_renderer";
+
+const DISPLAY_TYPES = {
+    NOTE: "line_note",
+    RECORD: "record",
+    ROOT: Symbol("root"),
+    SECTION: "line_section",
+    SUBSECTION: "line_subsection",
+};
+
+const SUB_SECTION_VALUE_FACTOR = 1;
+const TOP_SECTION_VALUE_FACTOR = 100;
+
+function isSection(type) {
+    return [DISPLAY_TYPES.SECTION, DISPLAY_TYPES.SUBSECTION].includes(type);
+}
+
+class HierarchyItem {
+    constructor(type, record, parent) {
+        this.type = type;
+        this.record = record;
+        this.parent = parent;
+        this.childItems = [];
+        this.childSections = [];
+    }
+
+    get displayName() {
+        return this.record.data.display_name || this.record.data.name;
+    }
+
+    get id() {
+        return this.record.id;
+    }
+
+    get index() {
+        const prop = this.isSection ? "childSections" : "childItems";
+        return this.parent[prop].indexOf(this) + 1;
+    }
+
+    get isLastChildItem() {
+        return this.index === this.parent.childItems.length;
+    }
+
+    get isNote() {
+        return this.type === DISPLAY_TYPES.NOTE;
+    }
+
+    get isRoot() {
+        return this.type === DISPLAY_TYPES.ROOT;
+    }
+
+    get isSection() {
+        return isSection(this.type);
+    }
+
+    get isTopSection() {
+        return this.type === DISPLAY_TYPES.SECTION;
+    }
+
+    get records() {
+        if (this.isRoot) {
+            return [
+                ...this.childItems.flatMap((it) => it.records),
+                ...this.childSections.flatMap((it) => it.records),
+            ];
+        }
+        if (this.isSection) {
+            return [
+                this.record,
+                ...this.childItems.flatMap((it) => it.records),
+                ...this.childSections.flatMap((it) => it.records),
+            ];
+        }
+        return [this.record];
+    }
+
+    get sectionValue() {
+        if (!this.isSection) {
+            return 0;
+        }
+        if (this.isTopSection) {
+            return this.index * TOP_SECTION_VALUE_FACTOR;
+        } else {
+            return this.parentValue + this.index * SUB_SECTION_VALUE_FACTOR;
+        }
+    }
+
+    addChild(item) {
+        const prop = item.isSection ? "childSections" : "childItems";
+        this[prop].push(item);
```

Comment:
nitpicking:
```suggestion
        if (item.isSection) {
            this.childSections.push(item);
        } else {
            this.childItems.push(item);
        }
```
easier to read, easier to grep (less chance to be missed by a grep).